mesh/mesh.go

```diff
@@ -500,3 +500,66 @@ func (m *Mesh) GetATXs(atxIds []types.AtxId) (map[types.AtxId]*types.ActivationT
 	}
 	return atxs, mIds
 }
+
+// ActiveSetForLayerView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (m *Mesh) ActiveSetForLayerView(layer types.LayerID, layersPerEpoch uint16) (uint32, error) {
+
+	epoch := layer.GetEpoch(layersPerEpoch)
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(layersPerEpoch)
+
+	// build a map of all blocks between firstLayerOfPrevEpoch and our current layer
+	mp := make(map[types.BlockID]struct{})
+	blocks, err := m.LayerBlockIds(layer)
+	for _, bid := range blocks {
+		mp[bid] = struct{}{}
+	}
+
+	activeSetSize := uint32(0)
+	countedAtxs := make(map[string]types.AtxId)
+
+	traversalFunc := func(blkh *types.BlockHeader) error {
+
+		// only contextually valid blocks
+		if !m.IsContexuallyValid(blkh.Id) {
+			return nil
+		}
+
+		blk, err := m.GetBlock(blkh.Id)
+		if err != nil {
+			return err
+		}
+
+		// count unique ATXs
+		for _, id := range blk.AtxIds {
+			atx, err := m.GetAtx(id)
+			if err != nil {
+				log.Panic("error fetching atx %v from database -- inconsistent state", id.ShortId()) // TODO: handle inconsistent state
+				return fmt.Errorf("error fetching atx %v from database -- inconsistent state", id.ShortId())
+			}
+
+			// make sure the target epoch is our epoch
+			if atx.TargetEpoch(layersPerEpoch) != epoch {
+				m.Debug("atx %v found, but targeting epoch %v instead of publication epoch %v",
+					atx.ShortId(), atx.TargetEpoch(layersPerEpoch), epoch)
+				continue
+			}
+			if prevId, exist := countedAtxs[atx.NodeId.Key]; exist { // same miner
+				if prevId != id { // different atx for same epoch
+					m.Error("Encountered second atx for the same miner on the same epoch atxId1=%v atxId2=%v", prevId, id)
+				}
+				continue
```

Comment:
hmm we have a problem here. In case a miner published two different ATXs you shouldn't count any of them. So the first time you detect that you need to mark the miner as invalid, reduce the atx count (only the first time!). Next time that you'll see this miner's ATX you'll ignore it since the miner is invalid.

Reply:
Fixed